tests/python_package_test/test_engine.py

```diff
@@ -1203,6 +1203,30 @@ def test_feature_name_with_non_ascii():
     assert feature_names == gbm2.feature_name()
 
 
+def test_parameters_are_loaded_from_model_file(tmp_path):
+    X = np.hstack([np.random.rand(100, 1), np.random.randint(0, 5, (100, 2))])
+    y = np.random.rand(100)
+    ds = lgb.Dataset(X, y)
+    params = {
+        'bagging_fraction': 0.8,
+        'bagging_freq': 2,
+        'boosting': 'rf',
+        'feature_contri': [0.5, 0.5, 0.5],
+        'feature_fraction': 0.7,
+        'boost_from_average': False,
+        'interaction_constraints': [[0, 1], [0]],
+        'metric': ['l2', 'rmse'],
+        'num_leaves': 5,
+        'num_threads': 1,
+    }
+    model_file = tmp_path / 'model.txt'
+    lgb.train(params, ds, num_boost_round=1, categorical_feature=[1, 2]).save_model(model_file)
+    bst = lgb.Booster(model_file=model_file)
+    set_params = {k: bst.params[k] for k in params.keys()}
+    assert set_params == params
+    assert bst.params['categorical_feature'] == [1, 2]
```

Comment:
These tests are checking that the `params` attribute on the Python `Booster` object matches what is in the file...but that being true doesn't necessarily guarantee that if you used this `Booster` for continued training that it would still actually use those parameters at the C++ side, right?

I'm bringing this up because I'm struggling to understand the relationship between `GBDT::loaded_parameter_`

https://github.com/microsoft/LightGBM/blob/9e89ee7fac7dfe18f0f4ab3ea7542854e3d0b936/src/boosting/gbdt_model_text.cpp#L600

and `GBDT::config_`

https://github.com/microsoft/LightGBM/blob/9e89ee7fac7dfe18f0f4ab3ea7542854e3d0b936/src/boosting/gbdt.h#L461-L462

In https://github.com/microsoft/LightGBM/issues/2613#issuecomment-955948360, I'd recommended using `GBDT::config_` as the source of the parameter information, because it seems to me like the parameters section of the model file is loaded up into that property `loaded_parameter_`, but not actually used to configure the `GBDT` object. But now I'm not sure that that's right either haha.

For example....if you tried continued training after loading a model from text file like this, would LightGBM respect `feature_fraction=0.7`? Or would it fall back to the default of `1.0`?

Reply:
> that being true doesn't necessarily guarantee that if you used this Booster for continued training that it would still actually use those parameters at the C++ side, right?

I believe the `loaded_parameter_` attribute was added in #1495 to have the model parameters stored, but it isn't actually used anywhere other than when writing the parameters back to a file when there's no config. https://github.com/microsoft/LightGBM/blob/d78b6bc2fdf96c87e3cb61f2d497a962e3270c91/src/boosting/gbdt_model_text.cpp#L393-L401 Not sure when this happens though, maybe from the CLI using refit task.

So I see these parameters more as "informative" of how the Booster was trained, rather than to be used to continue training. The original issue wanted to have `objective` be loaded back to be used in SHAP and I think having them may help trying to replicate previous trainings or having the `categorical_feature` for inference. I actually thought of maybe having an argument to decide whether or not to load them, they may not be useful and that computation could be skipped.

Reply:
@guolinke Could you please help to understand the connection between `config_` and  `loaded_parameter_`?

Reply:
IMO, I think the "loaded_parameter_"  is used to check the parameters used for the model file. For continued training, the current "config_" is used.

Reply:
> to check the parameters used for the model file

ok but @guolinke does `loaded_parameter_` actually get used to update `config_`?

If it does, then wrappers like the R and Python package shouldn't ever need to know about `loaded_parameter_`, and could just rely on `config_`.

If if doesn't, then I think it means that if you load a model from a text file and then perform continued training, you'll need to explicitly `.update_parameters()` (at least in the Python and R package), or otherwise the additional training rounds will just use all of the default values of LightGBM's parameters.

In my opinion, ensuring that a model loaded from a text file looks as similar as possible to one created in-memory by training (e.g. that both `Booster.params` in the Python object and the `config_` property on the C++ booster are identical to what they'd be after running `lgb.train()`) would simplify this feature and other features related to using models that were loaded from text files.

Reply:
@jameslamb I think the `config_` is used to update  `loaded_parameter_`.
User can still set new parameters when "loaded_parameter_", when continue training from a model. In this case, I think the new parameters have higher priority. 
So it is like, model_file -> loaed_parameters_ -> load_config -> update by user's new parameters.
The problem here is the default parameter values.

Reply:
@jmoralez @jameslamb any further comments about this?

Reply:
I added a warning about passing parameters to the booster constructor in 9467814 because in order to be able to call `Booster.update` you need a training set, and if you initialize a booster with the training set `lgb.Booster(model_file=..., train_set=...)` the `model_file` argument gets ignored, so I don't think there's a way to do incremental training based on the `Booster` object. Thus, I don't think it really matters what the parameters for the loaded booster are on the cpp side, because the incremental training would be through `lgb.train` that already supports overriding the parameters for the new iterations. WDYT?

Reply:
it sounds good to me

Reply:
Thanks @jmoralez 

It's an excellent point that `Booster`'s constructor has logic like

```python
if train_set is not None:
    ...
elif model_file is not None:
    ...
```

Meaning that you can't have both. I agree with the warning you've added and appreciate you adding a test for it as well!

I'm sorry for holding up this PR for so long over this. I'm still confused about the relationship between `config_` and `loaded_parameter_`, but I don't have any more concerns about this specific PR.

Reply:
I should also add, as I mentioned in https://github.com/microsoft/LightGBM/pull/5424#pullrequestreview-1092403723 ... I think we should merge this because future changes probably wouldn't require any user-facing breaking changes. The fact that the new function you've added in `c_api` takes in a `BoosterHandle` means that we could swap it to using a different property of the `Booster` (e.g. `config_` instead of `loaded_parameter_`) in the future without an API-level breaking change. Very nice!